src/main/scala/org/broadinstitute/dsde/workbench/sam/google/GoogleKeyCache.scala

```diff
@@ -0,0 +1,75 @@
+package org.broadinstitute.dsde.workbench.sam.google
+
+import java.io.ByteArrayInputStream
+
+import akka.http.scaladsl.model.{DateTime, StatusCodes}
+import cats.data.OptionT
+import cats.implicits._
+import com.google.api.client.googleapis.json.GoogleJsonResponseException
+import com.google.api.services.storage.model.StorageObject
+import org.broadinstitute.dsde.workbench.google.{GoogleIamDAO, GoogleStorageDAO}
+import org.broadinstitute.dsde.workbench.model.google.GoogleModelJsonSupport._
+import org.broadinstitute.dsde.workbench.model.google.{GoogleProject, ServiceAccountKey, ServiceAccountKeyId}
+import org.broadinstitute.dsde.workbench.model._
+import org.broadinstitute.dsde.workbench.sam.config.{GoogleServicesConfig, PetServiceAccountConfig}
+import org.broadinstitute.dsde.workbench.sam.service.KeyCache
+
+import scala.concurrent.{ExecutionContext, Future}
+import spray.json._
+
+/**
+  * Created by mbemis on 1/10/18.
+  */
+class GoogleKeyCache(val googleIamDAO: GoogleIamDAO, val googleStorageDAO: GoogleStorageDAO, val googleServicesConfig: GoogleServicesConfig, val petServiceAccountConfig: PetServiceAccountConfig)(implicit val executionContext: ExecutionContext) extends KeyCache {
+
+  override def onBoot(): Future[Unit] = {
+    googleStorageDAO.createBucket(googleServicesConfig.serviceAccountClientProject, petServiceAccountConfig.keyBucketName).map { _ =>
+      googleStorageDAO.setBucketLifecycle(petServiceAccountConfig.keyBucketName, petServiceAccountConfig.activeKeyMaxAge)
+    }
+  }
+
+  override def getKey(pet: PetServiceAccount, project: GoogleProject): Future[String] = {
+    val retrievedKeys = for {
+      keyObjects <- googleStorageDAO.listObjectsWithPrefix(petServiceAccountConfig.keyBucketName, keyNamePrefix(project, pet.serviceAccount.email))
+      keys <- googleIamDAO.listServiceAccountKeys(project, pet.serviceAccount.email)
+    } yield (keyObjects.toList, keys.toList)
+
+    retrievedKeys.flatMap {
+      case (Nil, _) => furnishNewKey(pet, project) //mismatch. there were no keys found in the bucket, but there may be keys on the service account
```

Comment:
there's a change i can make to the GoogleIamDAO that will make this case much less likely. it currently returns all system managed keys and user managed keys, but we can filter out the user managed keys